### PR TITLE
ZEPPELIN-598 ] Dynamic loading for Interpreter

### PR DESCRIPTION
**Pull request by cloverhearts**
### What is this PR for?
Use of external libraries that are included in the Roadmap,
The interpreter needs related to dynamic loading capabilities.

### What type of PR is it?
Feature

### Todos
* [x] - Dynamic load

### Is there a relevant Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-598
### How should this be tested?
By annotating the interpreter setting items below, try using the loadDynamicInterpreter method.
Or create a new interpreter loads created in the local Maven repository.
incubator-zeppelin/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java
``` java 
    ZEPPELIN_INTERPRETERS("zeppelin.interpreters", "org.apache.zeppelin.spark.SparkInterpreter,"
        + "org.apache.zeppelin.spark.PySparkInterpreter,"
        + "org.apache.zeppelin.spark.SparkSqlInterpreter,"
        + "org.apache.zeppelin.spark.DepInterpreter,"
        + "org.apache.zeppelin.markdown.Markdown,"
...
```
### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java [hunk 1]
```diff
@@ -69,24 +69,25 @@
   DependencyResolver depResolver;
 
   public InterpreterFactory(ZeppelinConfiguration conf,
-      AngularObjectRegistryListener angularObjectRegistryListener,
-      DependencyResolver depResolver)
+                            AngularObjectRegistryListener angularObjectRegistryListener,
+                            DependencyResolver depResolver)
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: Can you please change indention here and below back, to make sure it is consistent with the rest of the project code?

Here is the relevant style guide section [javaguide.html#s4.6.3-horizontal-alignment](https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment)

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java [hunk 2]
```diff
@@ -95,6 +96,96 @@ public InterpreterFactory(ZeppelinConfiguration conf, InterpreterOption defaultO
...
+                                        String intpClassName) {
+    return loadDynamicInterpreter(intpGroupName, intpName, artifact, intpClassName, null, false);
+  }
+
+  public boolean loadDynamicInterpreter(String intpGroupName, String intpName, String artifact,
+                                        String intpClassName, String repositoryUrl,
+                                        boolean isSnapShotRepo) {
+    String[] artifactItem = artifact.split(":");
+    String interpreterDesPath = conf.getString(ConfVars.ZEPPELIN_INTERPRETER_REPO_DIR);
+    interpreterDesPath += "/" + intpGroupName + "/" + intpName + "/";
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: Guava's `Joiner` or `String.format()` might come handy here and below.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java [hunk 3]
```diff
@@ -95,6 +96,96 @@ public InterpreterFactory(ZeppelinConfiguration conf, InterpreterOption defaultO
...
+
+    try {
+      if (repositoryUrl != null) {
+        depResolver.addRepo("dyInterpreterRepo", repositoryUrl, isSnapShotRepo);
+      }
+      logger.info("interpreter- path {}", interpreterLoadPath);
+      depResolver.load(artifact, interpreterDesPath);
+      setDynamicInterpreter(intpClassName, interpreterLoadPath);
+    } catch (Exception e) {
+      logger.error(e.getMessage());
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: Adding message like "Failed loading artifact ..." and passing in throwable using this logger's method  [error(java.lang.String, java.lang.Throwable)](http://www.slf4j.org/api/org/slf4j/Logger.html#error(java.lang.String, java.lang.Throwable)) might improve log message at runtime

**Comment by bzz** [comment]
Looks good to me, except for major style\formatting nit above.

**Comment by cloverhearts** [comment]
@bzz Thank you for feedback!
I will modify as you advise.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java [hunk 4]
```diff
@@ -95,6 +96,96 @@ public InterpreterFactory(ZeppelinConfiguration conf, InterpreterOption defaultO
...
+      return false;
+    }
+    return true;
+  }
+
+  public boolean unloadDynamicInterpreter(String intpGorupName, String intpName) {
+    try {
+      remove(intpName);
+    } catch (Exception e) {
+      logger.error(e.getMessage());
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: use `logger.error("BOOM!", e);` instead?
http://stackoverflow.com/questions/15805460/how-to-print-an-exception-using-logger

Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: also on line 127

**Comment by cloverhearts** [comment]
@bzz 
@felixcheung 

Thank you everyone.
I rebase to reflect and comment on all this PR, it has created a new PR.
The PR is'll close.

https://github.com/apache/incubator-zeppelin/pull/652

**Pull request closed without merging**